optimade/models/utils.py

```diff
@@ -1,8 +1,108 @@
+import inspect
+import warnings
 import re
+from enum import Enum
+from typing import Optional
+
+from pydantic import Field
+
+_PYDANTIC_FIELD_KWARGS = list(inspect.signature(Field).parameters.keys())
 
 __all__ = ("CHEMICAL_SYMBOLS", "EXTRA_SYMBOLS", "ATOMIC_NUMBERS", "SemanticVersion")
 
 
+class SupportLevel(Enum):
+    MUST = "must"
+    SHOULD = "should"
+    OPTIONAL = "optional"
+
+
+def StrictField(
+    *args,
+    description: str = None,
+    **kwargs,
+) -> Field:
+    """A wrapper around `pydantic.Field` that does the following:
+
+    - Forbids any "extra" keys that would be passed to `pydantic.Field`,
+      except those used elsewhere to modify the schema in-place,
+      e.g. "uniqueItems", "pattern" and those added by OptimadeField, e.g.
+      "unit", "queryable" and "sortable".
+    - Emits a warning when no description is provided.
+
+    Raises:
+        RuntimeError: if a Field is created with an unexpected key.
+
+    Returns:
+        the pydantic field
+
+    """
+
+    allowed_keys = [
+        "unit",
+        "pattern",
+        "uniqueItems",
+        "support",
+        "queryable",
+        "sortable",
+    ]
+    _banned = [k for k in kwargs if k not in set(_PYDANTIC_FIELD_KWARGS + allowed_keys)]
+
+    if _banned:
+        raise RuntimeError(
+            f"Not creating StrictField({args}, {kwargs}) with forbidden keywords {_banned}."
+        )
+
+    if description is not None:
+        kwargs["description"] = description
+
+    if description is None:
+        warnings.warn(
+            f"No description provided for StrictField specified by {args}, {kwargs}."
+        )
+
+    return Field(*args, **kwargs)
+
+
+def OptimadeField(
+    *args,
+    support: Optional[SupportLevel] = SupportLevel.MUST,
+    queryable: Optional[SupportLevel] = SupportLevel.MUST,
+    unit: Optional[str] = None,
+    **kwargs,
+) -> Field:
+    """A wrapper around `pydantic.Field` that adds OPTIMADE-specific
+    field paramters `queryable`, `support` and `unit`, indicating
+    the corresponding support level in the specification and the
+    physical unit of the field.
+
+    Keyword arguments:
+        support: the support level of the field itself, i.e. whether the field
+            can be null or omitted by an implementation.
+        queryable: the support level corresponding to the queryablility
+            of this field.
+        unit: a string describing the unit of the field.
```

Comment:
```suggestion
    Arguments:
        support: The support level of the field itself, i.e. whether the field
            can be null or omitted by an implementation.
        queryable: The support level corresponding to the queryablility
            of this field.
        unit: A string describing the unit of the field.
```

Reply:
These are all keyword arguments right? I think this is google docstring compliant?

Reply:
Not really, since you've written them explicitly as parameters for this function. Keyword arguments are whatever can be specified in `kwargs` (kwargs = `k`ey`w`ord `arg`ument`s`) :) At least, as far as I know.

Reply:
I guess we're arguing semantics here, but I would call any named argument a keyword argument, with `kwargs` just being a convention for capturing any extra named arguments. When writing docstrings I end up defaulting to the style expected by sphinx/napoleon (https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections) which uses `Parameters`/`Arguments`/`Args` and `Keyword arguments` as the section names for positional and named arguments respectively, so it depends what mkdocstrings is looking for... I notice that it doesn't seem to accept "Keyword arguments" in the existing validator docs https://www.optimade.org/optimade-python-tools/api_reference/validator/validator/

Reply:
Looks like mkdocstrings just lumps everything into "Arguments:" and distinguishes named arguments by their default values, so I guess we should go down that route